File: src/nova_audio_agent/executors/codex_jsonl.py
from __future__ import annotations

import json
from dataclasses import dataclass
from typing import Any, Literal, Mapping


class CodexProtocolError(RuntimeError):
    def __init__(self, code: str) -> None:
        super().__init__(code)
        self.code = code


@dataclass(frozen=True, slots=True)
class CodexProtocolSummary:
    events: tuple[Mapping[str, Any], ...]
    thread_started: bool
    turn_started: bool
    terminal: Literal["completed", "failed"] | None
    transport_closed: bool
    unknown_event_count: int
    internal_activity_count: int


class CodexJsonlParser:
    MAX_LINE_BYTES = 64 * 1024
    MAX_STDOUT_BYTES = 1024 * 1024

    _MILESTONES = {
        "thread.started",
        "turn.started",
        "turn.completed",
        "turn.failed",
        "error",
    }
    _ITEM_EVENTS = {"item.started", "item.updated", "item.completed"}

    def __init__(self) -> None:
        self._events: list[Mapping[str, Any]] = []
        self._thread_started = False
        self._turn_started = False
        self._terminal: Literal["completed", "failed"] | None = None
        self._transport_closed = False
        self._unknown_event_count = 0
        self._internal_activity_count = 0
        self._stdout_bytes = 0

    def feed(self, line: bytes) -> None:
        if self._transport_closed:
            raise CodexProtocolError("transport_closed")
        if len(line) > self.MAX_LINE_BYTES:
            raise CodexProtocolError("line_too_large")
        if self._stdout_bytes + len(line) > self.MAX_STDOUT_BYTES:
            raise CodexProtocolError("stdout_too_large")
        self._stdout_bytes += len(line)

        try:
            text = line.decode("utf-8")
            event = json.loads(text, parse_constant=_reject_json_constant)
        except (UnicodeDecodeError, ValueError, RecursionError):
            raise CodexProtocolError("malformed_jsonl") from None
        if not isinstance(event, dict) or not event:
            raise CodexProtocolError("invalid_event")

        event_type = event.get("type")
        if not isinstance(event_type, str):
            raise CodexProtocolError("invalid_event_type")

        if event_type in self._ITEM_EVENTS:
            if not self._turn_started or self._terminal is not None:
                raise CodexProtocolError("item_outside_turn")
            self._internal_activity_count += 1
            return
        if event_type not in self._MILESTONES:
            self._unknown_event_count += 1
            return

        if event_type == "thread.started":
            if self._thread_started:
                raise CodexProtocolError("duplicate_thread")
            if self._turn_started or self._terminal is not None:
                raise CodexProtocolError("thread_out_of_order")
            self._thread_started = True
        elif event_type == "turn.started":
            if not self._thread_started:
                raise CodexProtocolError("turn_before_thread")
            if self._turn_started or self._terminal is not None:
                raise CodexProtocolError("duplicate_turn")
            self._turn_started = True
        elif event_type == "turn.completed":
            self._require_active_turn_for_terminal()
            self._append_internal_activity()
            self._terminal = "completed"
        elif event_type == "turn.failed":
            self._require_active_turn_for_terminal()
            self._append_internal_activity()
            self._terminal = "failed"

        self._events.append({"type": event_type})

    def close(self) -> CodexProtocolSummary:
        if self._transport_closed:
            raise CodexProtocolError("transport_closed")
        self._transport_closed = True
        if self._terminal is None:
            raise CodexProtocolError("missing_terminal")
        return CodexProtocolSummary(
            events=tuple(self._events),
            thread_started=self._thread_started,
            turn_started=self._turn_started,
            terminal=self._terminal,
            transport_closed=self._transport_closed,
            unknown_event_count=self._unknown_event_count,
            internal_activity_count=self._internal_activity_count,
        )

    def _append_internal_activity(self) -> None:
        if self._internal_activity_count:
            self._events.append(
                {
                    "type": "internal_activity",
                    "count": self._internal_activity_count,
                }
            )

    def _require_active_turn_for_terminal(self) -> None:
        if self._terminal is not None:
            raise CodexProtocolError("duplicate_terminal")
        if not self._turn_started:
            raise CodexProtocolError("terminal_without_turn")


def _reject_json_constant(_value: str) -> None:
    raise ValueError("non-standard JSON constant")

File: runtime/src/codex-jsonl.ts
import {CodexProtocolError, MAX_INTERNAL_ACTIVITY} from './codex-protocol.js'

export const MAX_LINE_BYTES = 64 * 1024
export const MAX_STDOUT_BYTES = 1024 * 1024

const fatalDecoder = new TextDecoder('utf-8', {fatal: true})
const MILESTONES = new Set([
  'thread.started', 'turn.started', 'turn.completed', 'turn.failed', 'error',
])
const ITEM_EVENTS = new Set(['item.started', 'item.updated', 'item.completed'])

export class CodexJsonlProtocolError extends CodexProtocolError {
  constructor(code: string) {
    super(code)
    this.name = 'CodexJsonlProtocolError'
  }
}

export interface CodexJsonlSummary {
  readonly events: readonly Readonly<Record<string, unknown>>[]
  readonly thread_started: boolean
  readonly turn_started: boolean
  readonly terminal: 'completed' | 'failed' | null
  readonly transport_closed: boolean
  readonly unknown_event_count: number
  readonly internal_activity_count: number
}

export class CodexJsonlParser {
  static readonly MAX_LINE_BYTES = MAX_LINE_BYTES
  static readonly MAX_STDOUT_BYTES = MAX_STDOUT_BYTES

  readonly #events: Readonly<Record<string, unknown>>[] = []
  #threadStarted = false
  #turnStarted = false
  #terminal: 'completed' | 'failed' | null = null
  #transportClosed = false
  #unknownEventCount = 0
  #internalActivityCount = 0
  #stdoutBytes = 0

  feed(line: Uint8Array): void {
    if (this.#transportClosed) throw new CodexJsonlProtocolError('transport_closed')
    if (!(line instanceof Uint8Array)) throw new CodexJsonlProtocolError('malformed_jsonl')
    if (line.byteLength > MAX_LINE_BYTES) throw new CodexJsonlProtocolError('line_too_large')
    if (this.#stdoutBytes + line.byteLength > MAX_STDOUT_BYTES) {
      throw new CodexJsonlProtocolError('stdout_too_large')
    }
    this.#stdoutBytes += line.byteLength
    let event: unknown
    try {
      event = JSON.parse(fatalDecoder.decode(line)) as unknown
      assertSafeNumbers(event)
    } catch {
      throw new CodexJsonlProtocolError('malformed_jsonl')
    }
    if (!isPlainObject(event) || Object.keys(event).length === 0) {
      throw new CodexJsonlProtocolError('invalid_event')
    }
    const eventType = event.type
    if (typeof eventType !== 'string') throw new CodexJsonlProtocolError('invalid_event_type')
    if (ITEM_EVENTS.has(eventType)) {
      if (!this.#turnStarted || this.#terminal !== null) {
        throw new CodexJsonlProtocolError('item_outside_turn')
      }
      if (this.#internalActivityCount < MAX_INTERNAL_ACTIVITY) this.#internalActivityCount += 1
      return
    }
    if (!MILESTONES.has(eventType)) {
      if (this.#unknownEventCount < MAX_INTERNAL_ACTIVITY) this.#unknownEventCount += 1
      return
    }
    if (eventType === 'thread.started') {
      if (this.#threadStarted) throw new CodexJsonlProtocolError('duplicate_thread')
      if (this.#turnStarted || this.#terminal !== null) {
        throw new CodexJsonlProtocolError('thread_out_of_order')
      }
      this.#threadStarted = true
    } else if (eventType === 'turn.started') {
      if (!this.#threadStarted) throw new CodexJsonlProtocolError('turn_before_thread')
      if (this.#turnStarted || this.#terminal !== null) {
        throw new CodexJsonlProtocolError('duplicate_turn')
      }
      this.#turnStarted = true
    } else if (eventType === 'turn.completed') {
      this.#requireActiveTurnForTerminal()
      this.#appendInternalActivity()
      this.#terminal = 'completed'
    } else if (eventType === 'turn.failed') {
      this.#requireActiveTurnForTerminal()
      this.#appendInternalActivity()
      this.#terminal = 'failed'
    }
    this.#events.push(Object.freeze({type: eventType}))
  }

  close(): CodexJsonlSummary {
    if (this.#transportClosed) throw new CodexJsonlProtocolError('transport_closed')
    this.#transportClosed = true
    if (this.#terminal === null) throw new CodexJsonlProtocolError('missing_terminal')
    return Object.freeze({
      events: Object.freeze([...this.#events]),
      thread_started: this.#threadStarted,
      turn_started: this.#turnStarted,
      terminal: this.#terminal,
      transport_closed: true,
      unknown_event_count: this.#unknownEventCount,
      internal_activity_count: this.#internalActivityCount,
    })
  }

  #appendInternalActivity(): void {
    if (this.#internalActivityCount === 0) return
    this.#events.push(Object.freeze({
      type: 'internal_activity', count: this.#internalActivityCount,
    }))
  }

  #requireActiveTurnForTerminal(): void {
    if (this.#terminal !== null) throw new CodexJsonlProtocolError('duplicate_terminal')
    if (!this.#turnStarted) throw new CodexJsonlProtocolError('terminal_without_turn')
  }
}

function assertSafeNumbers(value: unknown): void {
  if (typeof value === 'number') {
    if (!Number.isFinite(value) || (Number.isInteger(value) && !Number.isSafeInteger(value))) {
      throw new TypeError('unsafe number')
    }
    return
  }
  if (Array.isArray(value)) {
    for (const item of value) assertSafeNumbers(item)
    return
  }
  if (isPlainObject(value)) {
    for (const item of Object.values(value)) assertSafeNumbers(item)
  }
}

function isPlainObject(value: unknown): value is Record<string, unknown> {
  if (typeof value !== 'object' || value === null || Array.isArray(value)) return false
  const prototype = Object.getPrototypeOf(value) as object | null
  return prototype === Object.prototype || prototype === null
}
